### DISCUSSION TITLE
Update song select to latest design

### DISCUSSION THREAD

**Discussion by mk56-spn** [ppy/osu#16638] [Q&A] [answered]
Frenzibyte recommended i open a discussion since i was doing this, Currently working on matching the figma design for the wedge , Q and A because frankly i dont know which fits better

**Comment by mk56-spn** [comment]
here is what ive managed so far, im aware that the lateral coloured wedge is missing the stars and should be layered behind the main container, i had to adjust the text size down from the 40 figma said to 30 to achieve the same size more or less, are the values not meant to line up?
![wedge](https://user-images.githubusercontent.com/74463310/151254150-6ded3a80-d171-410d-837b-1d497a16ea5b.png)

**Reply by bdach** [reply]
values may not necessarily line up due to scaling differences.

remaining items on comparing with design (https://www.figma.com/file/DXKwqZhD5yyb1igc3mKo1P/Song-Select-Draft-2?node-id=1%3A1276):
* the gradient effect on the background doesn't appear to match (should be horizontal not vertical)
* font weight and size on title and shadow on artist don't match
* no plays display below artist

**Reply by mk56-spn** [reply]
mhm, im not like trying to act like its anywhere near done just to clarify. but thanks ill adress those

**Reply by peppy** [reply]
Not sure what's up with that bar to the right but I don't think that's how the design intends it to look.

**Reply by mk56-spn** [reply]
> Not sure what's up with that bar to the right but I don't think that's how the design intends it to look.

Yeah. I'm still trying to understand how applying depth to a containers children works

**Reply by mk56-spn** [reply]
@peppy  seeing as the wedge is shrinking considerably im thinking offsetting the background towards the top is best, ive looked through like 100 ish mapsets and that would improve what gets shown on 70% of them just a small detail

**Comment by mk56-spn** [comment]
![sss](https://user-images.githubusercontent.com/74463310/151451655-6759a3b3-9c91-4634-9c90-833da07d5eea.png)
well im kinda stumped, ive arrived at errors on both remaining tasks, fixing the shear and fetching the plays, normally i wouldnt bother asking since it would be my fault but since these are both errors that only happen upon the game attempting to load the container im not sure how to proceed. a nudge in the right direction would be appreciated. https://github.com/mk56-spn/osu/tree/beatmap_info_wedge

**Reply by mk56-spn** [reply]
to clarify i understand the issue its getting a null value, but idk why i followed the same logic that is applied in beatmapcardnormal to fetch the playcount

**Reply by bdach** [reply]
it is not clear from the above screenshot what is trying to cache what. if you wish to receive help it's best to link the actual code you're working with

**Reply by mk56-spn** [reply]
> it is not clear from the above screenshot what is trying to cache what. if you wish to receive help it's best to link the actual code you're working with

i did link  the branch, or do you mean the specific part? and if so how would i best format it here to make what im referring to easily  readable

**Reply by bdach** [reply]
i missed the link somehow.

the issue that is you're trying to set a value to a `[Cached]` property (`WedgeInfoText.statisticsContainer`) too late - `load()`/BDL is too late. it needs to be initialised either in constructor, or in `CreateChildDependencies()` (if you decide to override that).

**Reply by mk56-spn** [reply]
alright thanks , i understand, how awful is the code looking?

**Comment by mk56-spn** [comment]
I ask you to bare with these noobish questions a little longer, in the hopes that itll play of for you guys long term.

so i implemented the ```OnlineViewContainer``` to try and avoid crashing osu when it cant fetch playcount yet im still  getting some slightly different null error, https://github.com/mk56-spn/osu/tree/beatmap_info_wedge i pushed to the branch
but if you dont wanna go thr
![boundingbox](https://user-images.githubusercontent.com/74463310/151663218-3990c650-d358-4286-a7e7-53ffe7b31faf.png)
ough heres the  screenshot of the main issues and related code
![info](https://user-images.githubusercontent.com/74463310/151663223-b7e13bab-ebfa-431c-b9ea-564ccc277f84.png)

**Reply by peppy** [reply]
Please post text, not unreadable screenshots, if you want help.

**Reply by bdach** [reply]
the above screenshots don't even show the real error. the real error is in the actual run logs and is:

```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at osu.Game.Beatmaps.Drawables.Cards.Statistics.PlayCountStatistic..ctor(IBeatmapSetOnlineInfo onlineInfo) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Beatmaps/Drawables/Cards/Statistics/PlayCountStatistic.cs:line 19
   at osu.Game.Screens.Select.BeatmapInfoWedge.WedgeInfoText.createStatistics()+MoveNext() in /home/dachb/Dokumenty/opensource/osu/osu.Game/Screens/Select/BeatmapInfoWedge.cs:line 289
   at osu.Framework.Graphics.Containers.Container`1.AddRange(IEnumerable`1 range)
   at osu.Framework.Graphics.Containers.Container`1.set_ChildrenEnumerable(IEnumerable`1 value)
   at osu.Game.Screens.Select.BeatmapInfoWedge.WedgeInfoText.load(OsuColour colours, LocalisationManager localisation, BeatmapDifficultyCache difficultyCache) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Screens/Select/BeatmapInfoWedge.cs:line 204
```

and the cause is:

![2022-01-29-145941_793x220_scrot](https://user-images.githubusercontent.com/20418176/151663906-b2d26616-840e-4158-a9a2-05e553161a08.png)

figuring out the rest is left as an exercise for the reader.

i'm sorry but i cannot personally continue to provide help on such a basic level.

**Reply by bdach** [reply]
not to mention the fact that using both components you're using for that (`BeatmapCardStatistic` and `OnlineViewContainer`) is most likely not going to be correct. in the case of the wedge, if the user is offline playcounts should most likely just not show at all rather than print "not available" to the screen.

**Reply by mk56-spn** [reply]
uhhh the screenshots are pretty high res so i thought you would be able to see them well, my bad, the issue is im  trying to fetch ```createstatistics``` through an ```onlineViewContainer``` and its doesnt seem  to be doing its job of hiding itself when it doesnt receive information (or atleast thats what ive understood) 

here is the relevant code excerpt
``` new FillFlowContainer
                    {
                        Name = "Centre-aligned metadata",
                        Anchor = Anchor.CentreLeft,
                        Origin = Anchor.TopLeft,
                        Y = -45,
                        Direction = FillDirection.Vertical,
                        Padding = new MarginPadding { Left = 64, Right = shear_width * 3.5f },
                        AutoSizeAxes = Axes.Y,
                        RelativeSizeAxes = Axes.X,
                        Children = new Drawable[]
                        {
                            TitleLabel = new OsuSpriteText
                            {
                                Font = OsuFont.TorusAlternate.With(size: 30, weight: FontWeight.SemiBold),
                                RelativeSizeAxes = Axes.X,
                                Truncate = true,
                            },
                            ArtistLabel = new OsuSpriteText
                            {
                                Font = OsuFont.Torus.With(size: 20, italics: true),
                                RelativeSizeAxes = Axes.X,
                                Truncate = true,
                            },

                            statisticsContainer = new OnlineViewContainer ("Not available")
                            {
                                RelativeSizeAxes = Axes.X,
                                AlwaysPresent = true,
                                ChildrenEnumerable = createStatistics()
                            },
                        }
                    }
                };

                titleBinding.BindValueChanged(_ => setMetadata(metadata.Source));
                artistBinding.BindValueChanged(_ => setMetadata(metadata.Source), true);

                starDifficulty = difficultyCache.GetBindableDifficulty(beatmapInfo, (cancellationSource = new CancellationTokenSource()).Token);
                starDifficulty.BindValueChanged(s =>
                {
                    starRatingDisplay.FadeIn(transition_duration);
                    starRatingDisplay.Current.Value = s.NewValue ?? default;
                });

                starDifficulty = difficultyCache.GetBindableDifficulty(beatmapInfo, (cancellationSource = new CancellationTokenSource()).Token);
            }

            private IEnumerable<BeatmapCardStatistic> createStatistics()
            {
                yield return new PlayCountStatistic(BeatmapSet);
            }```

**Reply by mk56-spn** [reply]
right understandable, ill try and figure it out alone

**Comment by Gabixel** [comment, marked as answer]
I don't know if this discussion is still active/checked, but I guess it can be transferred to #18890 (which will keep track of this)?
